File: marker/marker.go
// Package marker implements the seimark marker body defined in docs/format.md.
package marker

import (
	"encoding/binary"
	"errors"
	"fmt"
	"time"
)

// Version is the only marker body version this package encodes and decodes.
const (
	Version          = 1
	FixedSize        = 22
	PayloadSoftLimit = 4096
	PayloadHardLimit = 65535
)

// FormatUUID identifies a seimark marker inside a user_data_unregistered SEI message.
func FormatUUID() [16]byte {
	return [16]byte{0x44, 0xa7, 0x3c, 0xb9, 0xb3, 0x6c, 0x45, 0x9a, 0x8f, 0x1a, 0xa3, 0xaa, 0x43, 0x1f, 0x62, 0x4a}
}

// TimeSource says which event OriginTime records.
type TimeSource uint8

// TimeSend means OriginTime is when the marker was sent.
// TimeCapture means OriginTime is when the frame was captured.
const (
	TimeSend    TimeSource = 0
	TimeCapture TimeSource = 1
)

func (s TimeSource) String() string {
	switch s {
	case TimeSend:
		return "send"
	case TimeCapture:
		return "capture"
	}
	return fmt.Sprintf("time_source(%d)", uint8(s))
}

// Marker is one decoded marker. Payload is nil when the marker carries none and
// non-nil, possibly empty, when the payload flag is set.
type Marker struct {
	TimeSource TimeSource
	OriginTime time.Time
	Sequence   uint32
	StreamID   [8]byte
	Payload    []byte
}

// ErrUnsupportedVersion is returned when a marker body's version byte is not Version.
var (
	ErrUnsupportedVersion = errors.New("seimark: unsupported marker version")
	ErrTruncated          = errors.New("seimark: truncated marker body")
	ErrPayloadTooLarge    = errors.New("seimark: payload exceeds 65535 bytes")
)

const (
	flagTimeCapture = 1 << 0
	flagPayload     = 1 << 1
)

// Decode parses a marker body. Reserved flag bits and bytes after the body are ignored.
func Decode(body []byte) (Marker, error) {
	if len(body) < FixedSize {
		return Marker{}, fmt.Errorf("%w: %d bytes, need %d", ErrTruncated, len(body), FixedSize)
	}
	if body[0] != Version {
		return Marker{}, fmt.Errorf("%w: %d", ErrUnsupportedVersion, body[0])
	}
	flags := body[1]
	m := Marker{
		OriginTime: time.UnixMicro(readInt64(body[2:10])).UTC(),
		Sequence:   binary.BigEndian.Uint32(body[10:14]),
	}
	if flags&flagTimeCapture != 0 {
		m.TimeSource = TimeCapture
	}
	copy(m.StreamID[:], body[14:22])
	if flags&flagPayload != 0 {
		if len(body) < FixedSize+2 {
			return Marker{}, fmt.Errorf("%w: payload flag set but no length", ErrTruncated)
		}
		n := int(binary.BigEndian.Uint16(body[22:24]))
		if len(body) < FixedSize+2+n {
			return Marker{}, fmt.Errorf("%w: payload length %d, %d bytes left", ErrTruncated, n, len(body)-FixedSize-2)
		}
		m.Payload = append([]byte{}, body[24:24+n]...)
	}
	return m, nil
}

// readInt64 reads eight big-endian bytes; docs/format.md defines the origin
// time as a signed 64-bit value, so the two's-complement reinterpretation is
// the specified conversion and cannot overflow.
func readInt64(b []byte) int64 {
	//nolint:gosec // G115: docs/format.md defines the origin time as signed 64-bit; this is the two's-complement read.
	return int64(binary.BigEndian.Uint64(b))
}

// IsFormatUUID reports whether uuid is the seimark format UUID.
func IsFormatUUID(uuid []byte) bool {
	return len(uuid) == 16 && [16]byte(uuid) == FormatUUID()
}

// Encode returns the marker body. The payload flag is set when Payload is non-nil.
// OriginTime is rounded to microseconds.
func (m Marker) Encode() ([]byte, error) {
	if len(m.Payload) > PayloadHardLimit {
		return nil, fmt.Errorf("%w: %d", ErrPayloadTooLarge, len(m.Payload))
	}
	out := make([]byte, FixedSize, FixedSize+2+len(m.Payload))
	out[0] = Version
	if m.TimeSource == TimeCapture {
		out[1] |= flagTimeCapture
	}
	binary.BigEndian.PutUint64(out[2:10], uint64(m.OriginTime.UnixMicro()))
	binary.BigEndian.PutUint32(out[10:14], m.Sequence)
	copy(out[14:22], m.StreamID[:])
	if m.Payload != nil {
		n := len(m.Payload)
		if n > PayloadHardLimit {
			return nil, fmt.Errorf("%w: %d", ErrPayloadTooLarge, n)
		}
		out[1] |= flagPayload
		out = binary.BigEndian.AppendUint16(out, uint16(n))
		out = append(out, m.Payload...)
	}
	return out, nil
}

File: browser/src/marker.ts
import { SeimarkError } from './errors.ts';

/** The marker format version this implementation writes and is the only one it reads. */
export const VERSION = 1;
export const FIXED_SIZE = 22;
/**
 * Payload size above which a mark still succeeds but reports a
 * `payload_above_soft_limit` warning, since a large payload is carried on every
 * marked frame.
 */
export const PAYLOAD_SOFT_LIMIT = 4096;
/** Payload size above which encoding is refused: the length field is 16 bits. */
export const PAYLOAD_HARD_LIMIT = 65535;
export const UUID_SIZE = 16;
/** Required length in bytes of a stream id; any other length is an `invalid_argument`. */
export const STREAM_ID_SIZE = 8;

/**
 * The 16-byte UUID that identifies a seimark message inside a
 * `user_data_unregistered` SEI, distinguishing it from any other vendor's.
 */
export const FORMAT_UUID: Readonly<Uint8Array> = Uint8Array.of(
  0x44, 0xa7, 0x3c, 0xb9, 0xb3, 0x6c, 0x45, 0x9a,
  0x8f, 0x1a, 0xa3, 0xaa, 0x43, 0x1f, 0x62, 0x4a,
);

/**
 * Which clock reading the origin time is: `'capture'` when the frame carried a
 * capture time, otherwise `'send'`, taken as the frame is encoded.
 */
export type TimeSource = 'send' | 'capture';

/** One decoded marker: the per-frame metadata carried in a frame's own SEI. */
export interface Marker {
  /** Which clock `originTimeUs` came from. */
  timeSource: TimeSource;
  /**
   * Origin time in microseconds since the Unix epoch. A `bigint` because the
   * field is a signed 64-bit count and a `number` would lose precision at the
   * top of the range.
   */
  originTimeUs: bigint;
  /** Per-stream counter, incremented once per marked frame, wrapping at 2^32. */
  sequence: number;
  /** Eight bytes identifying the source; see `STREAM_ID_SIZE`. */
  streamId: Uint8Array;
  /** Application bytes the sender attached, or null when the payload flag is clear. */
  payload: Uint8Array | null;
}

const OFF_FLAGS = 1;
const OFF_TIME = 2;
const OFF_SEQUENCE = 10;
const OFF_STREAM_ID = 14;
const OFF_PAYLOAD_LEN = 22;
const FLAG_CAPTURE = 0b01;
const FLAG_PAYLOAD = 0b10;

const INT64_MIN = -(2n ** 63n);
const INT64_MAX = 2n ** 63n - 1n;

/** Reports whether these bytes are the seimark UUID, so a foreign SEI is left alone. */
export function isFormatUUID(uuid: Uint8Array): boolean {
  return uuid.length === UUID_SIZE && FORMAT_UUID.every((b, i) => uuid[i] === b);
}

/**
 * Decodes a marker body, the SEI payload with the UUID already stripped. Throws
 * `SeimarkError` with `truncated` if the body is short and
 * `unsupported_version` if it is not version 1.
 */
export function decodeMarker(body: Uint8Array): Marker {
  if (body.length < FIXED_SIZE) {
    throw new SeimarkError('truncated', `marker body is ${body.length} bytes, needs ${FIXED_SIZE}`);
  }

  const version = body[0]!;
  if (version !== VERSION) {
    throw new SeimarkError('unsupported_version', `marker version ${version}`);
  }

  const view = new DataView(body.buffer, body.byteOffset, body.byteLength);
  const flags = body[OFF_FLAGS]!;

  let payload: Uint8Array | null = null;
  if ((flags & FLAG_PAYLOAD) !== 0) {
    if (body.length < OFF_PAYLOAD_LEN + 2) {
      throw new SeimarkError('truncated', 'payload flag set without a length');
    }
    const n = view.getUint16(OFF_PAYLOAD_LEN);
    const start = OFF_PAYLOAD_LEN + 2;
    if (body.length < start + n) {
      throw new SeimarkError('truncated', `payload length ${n} overruns the body`);
    }
    payload = body.slice(start, start + n);
  }

  return {
    timeSource: (flags & FLAG_CAPTURE) !== 0 ? 'capture' : 'send',
    originTimeUs: view.getBigInt64(OFF_TIME),
    sequence: view.getUint32(OFF_SEQUENCE),
    streamId: body.slice(OFF_STREAM_ID, OFF_STREAM_ID + STREAM_ID_SIZE),
    payload,
  };
}

/**
 * Encodes a marker body, without the UUID or the SEI wrapper. Throws
 * `SeimarkError` with `invalid_argument` if the stream id is not
 * `STREAM_ID_SIZE` bytes or the origin time is outside the signed 64-bit range,
 * and `payload_too_large` above `PAYLOAD_HARD_LIMIT`.
 */
export function encodeMarker(m: Marker): Uint8Array {
  if (m.streamId.length !== STREAM_ID_SIZE) {
    throw new SeimarkError('invalid_argument', `stream id is ${m.streamId.length} bytes, needs ${STREAM_ID_SIZE}`);
  }
  if (m.originTimeUs < INT64_MIN || m.originTimeUs > INT64_MAX) {
    throw new SeimarkError('invalid_argument', 'origin time is outside the signed 64-bit range');
  }
  if (m.payload !== null && m.payload.length > PAYLOAD_HARD_LIMIT) {
    throw new SeimarkError('payload_too_large', `payload is ${m.payload.length} bytes`);
  }

  const size = m.payload === null ? FIXED_SIZE : OFF_PAYLOAD_LEN + 2 + m.payload.length;
  const out = new Uint8Array(size);
  const view = new DataView(out.buffer);

  out[0] = VERSION;
  out[OFF_FLAGS] =
    (m.timeSource === 'capture' ? FLAG_CAPTURE : 0) | (m.payload === null ? 0 : FLAG_PAYLOAD);
  view.setBigInt64(OFF_TIME, m.originTimeUs);
  view.setUint32(OFF_SEQUENCE, m.sequence >>> 0);
  out.set(m.streamId, OFF_STREAM_ID);

  if (m.payload !== null) {
    view.setUint16(OFF_PAYLOAD_LEN, m.payload.length);
    out.set(m.payload, OFF_PAYLOAD_LEN + 2);
  }

  return out;
}
